tests/unit/subscribeToPropertiesTest.js

```diff
@@ -0,0 +1,70 @@
+import Onyx from '../../lib';
+import waitForPromisesToResolve from '../utils/waitForPromisesToResolve';
+
+const ONYX_KEYS = {
+    TEST_KEY: 'test',
+    COLLECTION: {
+        TEST_KEY: 'test_',
+    },
+};
+
+Onyx.init({
+    keys: ONYX_KEYS,
+    registerStorageEventListener: () => {},
+    initialKeyStates: {
+        [ONYX_KEYS.OTHER_TEST]: 42,
+    },
+});
+
+describe('Onyx property subscribers', () => {
+    let connectionID;
+
+    afterEach(() => {
+        Onyx.disconnect(connectionID);
+        return Onyx.clear();
+    });
+
+    it('Triggers a connection callback when the property changes', () => {
+        const connectionCallbackMock = jest.fn();
+
+        return waitForPromisesToResolve()
+
+            // Given that onyx contains an object with two properties
+            .then(() => Onyx.set(ONYX_KEYS.TEST_KEY, {a: 'one', b: 'two'}))
+
+            // When we connect to that key and specific a selector path
+            .then(() => {
+                connectionID = Onyx.connect({
+                    key: ONYX_KEYS.TEST_KEY,
+                    selector: 'a',
+                    callback: connectionCallbackMock,
+                });
+                return waitForPromisesToResolve();
+            })
+
+            .then(() => {
+                // Then the callback should be called once
+                expect(connectionCallbackMock).toHaveBeenCalledTimes(1);
+
+                // with the value of just the .a value
+                expect(connectionCallbackMock).toHaveBeenCalledWith('one', ONYX_KEYS.TEST_KEY);
```

Comment:
I'm not really sure what we want the returned value to be.

1. The value of the selector path (this is what I've implemented)
2. An object containing _only_ the properties that were selected (in this case, it would return `{a:'one'}`)

I think `2` is what @marcaaron was kind of going for in his original proposal, but it wasn't clear to me why that would be preferred. 

I'm happy to do it either way. Whichever way we think will be the most intuitive.

Reply:
Also, if we go for option `2`, how does that work with arrays?

```
data = [1, 2, 3];
selector = '[1]';
```

Should that return `2` or `[2]`?

Reply:
Mentioned on the issue but I was thinking that property mapping would be useful and in particular when using `withOnyx()`.

I think since we don't have any property mapping for `Onyx.connect()` today that a simple path selector like this should just return the value. 

That said, I think I could see a function working pretty well if you _did_ want some properties mapped to an object. I would have to look at some examples to see if that's something we want or not - but feels like it could be.

```js
Onyx.connect({
    key: 'testObject',
    callback: (myMappedObject) => {...},
    selector(testObject) {
        const myMappedObject = {
            propertyA: testObject.propertyA,
            propertyB: testObject.propertyB,
        };
        return myMappedObject;
    }
})
```

Reply:
Oh, that's a really interesting idea to have `selector` as a mapping function.

Reply:
> I'm not really sure what we want the returned value to be.

IMO 1, just the value you asked.

> Also, if we go for option 2, how does that work with arrays?

I don't get the example, if you wanted the first element of the array, then selector would be `0` and it would return `1`

> but feels like it could be.

This sounds good, but not sure if we need it. In any case I think we should still support the "easy" case of string selectors.

Reply:
Ok then my next question is **how are we actually going to use this feature** because we can just look and see how we would use it and then go with a design that makes sense. 

If we are going to want to subscribe to multiple properties of an object in many places then we will be adding a new subscriber for each one. That doesn't really seem as efficient - but maybe we don't want or need to subscribe to multiple properties. Do we have a clear idea that it is more useful to subscribe to one property of an object at a time? Because that's what we are optimizing for now.

Reply:
Agree. There's at least one place I know we will want to subscribe to 2 fields: [here](https://github.com/Expensify/App/blob/e76c0e4e646c10f7e143c773ecc580e25634b8d0/src/components/AvatarWithIndicator.js#L78).

But that can also be solved by allowing an array of strings in `selector`.

I also assume that some places using the OfflineWithFeedback component might also need 2 (since it needs to read the error and the pendingAction), but it's possible that in all those cases, we need the full object anyway (to use it in the rest of the component).